rmw_zenoh_cpp/src/detail/rmw_data_types.hpp

```diff
@@ -70,9 +71,10 @@ struct rmw_node_data_t
   zc_owned_liveliness_token_t token;
 };
 
-///==============================================================================
-struct rmw_publisher_data_t
+///=============================================================================
+class rmw_publisher_data_t : public EventsBase
```

Comment:
I have to say that making all of publishers, subscriptions, services, and clients derive from `EventsBase` seems weird.  That is, inheritance is an "is-a" relationship, and I don't think of publishers as events, for instance.

As I understand it, the basic problem here is that we want to be able to call `trigger_event`, `event_set_callback`, `pop_next_event`, `event_queue_is_empty`, `event_attach_condition`, and `event_detach_condition` on all of these.    One way I can think to do this is to make another class called `RMWEntity`, or something like that, and then `RMWEntity` would contain an EventBase class.  That way it would be a "has-a" relationship.  There may also be some other ways of restructuring this class hierarchy to make it clearer what we are trying to do; I think it is worth thinking about some more and exploring.

Reply:
Thanks for the suggestion. Happy to switch to the "has-a" pattern. How about a restructuring like this:
* I will split the `EntitiesBase` into `UserCallbackManager` and `QoSEventManager` classes. The former handles user specified callbacks while the later manages QoS events that we support.
* The `rmw_subscription/client/service_data_t` classes will have an object of type `CallbackManager` while only `rmw_publisher/subscription_data_t` classes will additionally contain `QoSEventsManger` objects since QoS events are not are not something for clients & services.

```bash
class UserCallbackManager
{
public:
  /// @brief Set the user defined callback that should be called when
  /// a new message/response/request is received.
  /// @param user_data the data that should be passed to the callback.
  /// @param callback the callback to be set.
  void set_user_callback(const void * user_data, rmw_event_callback_t callback);

  /// Trigger the user callback.
  void trigger_user_callback();

  
private:
  std::mutex event_mutex_;
  /// User callback that can be set via set_user_callback().
  rmw_event_callback_t callback_ {nullptr};
  /// User data that should be passed to the user callback.
  const void * user_data_ {nullptr};
  /// number of trigger requests made before the callback was set.
  size_t unread_count_ {0};
};

class QoSEventsManager
{
public:
  /// @brief  Set the callback to be triggered when the relevant event is triggered.
  /// @param event_id the id of the event
  /// @param callback the callback to trigger for this event.
  /// @param user_data the data to be passed to the callback.
  void event_set_callback(
    rmw_zenoh_event_type_t event_id,
    rmw_event_callback_t callback,
    const void * user_data);

  /// @brief  Returns true if the event queue is empty.
  /// @param event_id the event id whose event queue should be checked.
  bool event_queue_is_empty(rmw_zenoh_event_type_t event_id) const;

  /// Pop the next event in the queue.
  /// @param event_id the event id whose queue should be popped.
  std::unique_ptr<rmw_zenoh_event_status_t> pop_next_event(
    rmw_zenoh_event_type_t event_id);

  /// Add an event status for an event.
  /// @param event_id the event id queue to which the status should be added.
  void add_new_event(
    rmw_zenoh_event_type_t event_id,
    std::unique_ptr<rmw_zenoh_event_status_t> event);

  /// @brief Attach the condition variable provided by rmw_wait.
  /// @param condition_variable to attach.
  void attach_event_condition(
    rmw_zenoh_event_type_t event_id,
    std::condition_variable * condition_variable);

  /// @brief Detach the condition variable provided by rmw_wait.
  void detach_event_condition(rmw_zenoh_event_type_t event_id);

private:
  /// @brief Trigger the callback for an event.
  /// @param event_id the event id whose callback should be triggered.
  void trigger_event_callback(rmw_zenoh_event_type_t event_id);

  /// Notify once event is added to an event queue.
  void notify_event(rmw_zenoh_event_type_t event_id);

  /// Mutex to lock when read/writing members.
  // The mutex is recursive as add_new_event() invokes `trigger_event_callback()`.
  mutable std::recursive_mutex event_mutex_;
  /// Mutex to lock for event_condition.
  mutable std::mutex event_condition_mutex_;
  /// Condition variable to attach for event notifications.
  std::condition_variable * event_conditions_[ZENOH_EVENT_ID_MAX + 1]{nullptr};
  rmw_event_callback_t event_callback_[ZENOH_EVENT_ID_MAX + 1] {nullptr};
  const void * event_data_[ZENOH_EVENT_ID_MAX + 1] {nullptr};
  size_t event_unread_count_[ZENOH_EVENT_ID_MAX + 1] {0};
  // A dequeue of events for each type of event this RMW supports.
  std::deque<std::unique_ptr<rmw_zenoh_event_status_t>> event_queues_[ZENOH_EVENT_ID_MAX + 1] {};
  const std::size_t event_queue_depth_ = 10;
};
```

Reply:
No more inheritance :) https://github.com/ros2/rmw_zenoh/pull/103/commits/29f3a79f039ef679fdf9a75e6cd233d6445ae7aa